### PR TITLE
[PFS] GraphQL Client

### PR DESCRIPTION
**Pull request by roopakv**
# Description
This PR introduces the graphQL client for M2. This will be used to fetch feature switches from Bolt.

@mikea could you take a look at the way I structured the query and let me know what you think?

#changelog Adds a graphQL client to talk to Bolt.

Fixes: https://app.asana.com/0/1141250355664307/1144434411596971

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [x] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes
- [x] I have created or modified unit tests to sufficiently cover my changes.

### CODE REVIEW

**Comment by matt-thomason** [comment]
@roopakv Please link asana task.

**Comment by roopakv** [comment]
@matt-thomason done

File: Helper/GraphQL/Client.php [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        $this->responseFactory = $responseFactory;
+        $this->requestFactory = $requestFactory;
+        $this->logHelper = $logHelper;
+        $this->bugsnag = $bugsnag;
+    }
+
+    private function makeGQLCall($query, $operation, $variables) {
+        $result = $this->responseFactory->create();
+        $client = $this->httpClientFactory->create();
+        error_log("KOKOKOKOK");
```
Comment (daisy1754) [inline Helper/GraphQL/Client.php]: remove

File: Helper/GraphQL/Client.php [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+        $apiKey = $this->configHelper->getApiKey();
+
+        $gql_request = array(
```
Comment (daisy1754) [inline Helper/GraphQL/Client.php]: nit - let's stick to camelCase $gqlRequest

File: Helper/GraphQL/Client.php [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+            "operationName" => $operation,
+            "variables" => $variables,
+            "query" => $query
+        );
+
+        $requestData = json_encode($gql_request, JSON_UNESCAPED_SLASHES);
+
+        $apiURL = $this->configHelper->getApiUrl() . self::MERCHANT_API_GQL_ENDPOINT;
+        error_log($apiURL);
```
Comment (daisy1754) [inline Helper/GraphQL/Client.php]: remove

File: Helper/GraphQL/Client.php [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+        $client->setUri($apiURL);
+
+        $client->setConfig(['maxredirects' => 0, 'timeout' => 30]);
+
+        $headers =  [
```
Comment (daisy1754) [inline Helper/GraphQL/Client.php]: Hm, can we somehow share this with `Helper/Api`?

File: Helper/Shared/Utils.php [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+ * @license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+
+namespace Bolt\Boltpay\Helper\Shared;
+
+
+use Magento\Framework\Exception\LocalizedException;
+
+
+class Utils {
```
Comment (daisy1754) [inline Helper/Shared/Utils.php]: how about `ApiUtils` so it's little more specific?

File: Helper/GraphQL/Constants.php [hunk 6]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+
+namespace Bolt\Boltpay\Helper\GraphQL;
+
+
+class Constants {
```
Comment (daisy1754) [inline Helper/GraphQL/Constants.php]: RequestConstants? I prefer name to be specific so this class won't be dump all

Comment (roopakv) [reply Helper/GraphQL/Constants.php]: its already within GQL namespace right? i want to put all GQL constants here

File: Helper/GraphQL/Client.php (hunk 1 above)
Comment (roopakv) [reply Helper/GraphQL/Client.php]: oops

File: Helper/GraphQL/Client.php (hunk 4 above)
Comment (roopakv) [reply Helper/GraphQL/Client.php]: yeah let me move to a constants class

File: Helper/Shared/Utils.php (hunk 5 above)
Comment (roopakv) [reply Helper/Shared/Utils.php]: Love it

File: Helper/GraphQL/Client.php [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+                        'bolt_trace_id' => @$headers[ConfigHelper::BOLT_TRACE_ID_HEADER],
+                    ]
+                ]);
+            });
+        } catch (\Exception $e) {
+            throw new LocalizedException($this->wrapGatewayError($e->getMessage()));
+        }
+
+        if ($responseBody) {
+            error_log("ok got bdy");
```
Comment (mikea) [inline Helper/GraphQL/Client.php]: cleanup

File: Helper/GraphQL/Constants.php [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+    // The const via which Bolt identifies the type of plugin.
+    const PLUGIN_TYPE='MAGENTO2';
+
+    /**
+     * The graphql query to retrieve feature switches. This will be maintained backward compatible.
+     */
+    const GET_FEATURE_SWITCHES_QUERY = <<<'GQL'
+query GetFeatureSwitches($type: PluginType!, $version: String!) {
```
Comment (mikea) [inline Helper/GraphQL/Constants.php]: this is exactly how we have it in storm!

**Comment by roopakv** [comment]
@daisy1754 is there a doc about running tests locally?

**Comment by daisy1754** [comment]
@daisy1754 https://github.com/BoltApp/bolt-magento2/blob/master/Test/Unit/README.md is the only doc AFAIK

**Comment by danac-gs** [comment]
@roopakv https://github.com/BoltApp/bolt-magento2/blob/master/Test/Unit/README.md
However you might need to tweak the config a bit (if the plugin files are in vendor/boltpay/bolt-magento2 and not in app/code/Bolt/Boltpay). Here's mine:
```xml
<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:noNamespaceSchemaLocation="http://schema.phpunit.de/6.2/phpunit.xsd"
         colors="true"
         convertErrorsToExceptions="true"
         convertNoticesToExceptions="true"
         convertWarningsToExceptions="true"
         strict="true"
         bootstrap="./framework/bootstrap.php"
>
    <testsuite name="Bolt Unit Tests">
        <directory suffix="Test.php">../../../vendor/boltpay/bolt-magento2/Test/Unit</directory>
    </testsuite>
    <php>
        <ini name="date.timezone" value="America/Los_Angeles"/>
        <ini name="xdebug.max_nesting_level" value="300"/>
    </php>
    <filter>
        <whitelist addUncoveredFilesFromWhiteList="true">
            <directory suffix=".php">../../../vendor/boltpay/bolt-magento2</directory>
            <exclude>
                <directory>../../../vendor/boltpay/bolt-magento2/Test</directory>
            </exclude>
        </whitelist>
    </filter>
    <listeners>
        <listener class="Magento\Framework\TestFramework\Unit\Listener\ReplaceObjectManager"/>
    </listeners>
    <logging>
        <log type="coverage-text" target="php://stdout" showUncoveredFiles="false"/>
    </logging>
</phpunit>
```

**Comment by matt-thomason** [comment]
@roopakv I believe @ethanwayda22 wrote a small setup for running in phpstorm, but I think it also needs a bit of tweaking. https://www.notion.so/boltteam/Unit-Test-Guide-89f11aa8f1a34c44a8d08d5ca89f6128

File: Helper/GraphQL/Constants.php [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+      name
+      value
+      defaultValue
+      rolloutPercentage
+    }
+  }
+}
+GQL;
+
+    // Operation name for graphql retrieve ferature switch query.
```
Comment (matt-thomason) [inline Helper/GraphQL/Constants.php]: comment typo 😄

File: Helper/Shared/ApiUtils.php [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+    private static function isApiError($response) {
+        $arr = (array)$response;
+        return array_key_exists('errors', $arr) ||
+            array_key_exists('error_code', $arr) ||
+            array_key_exists('error', $arr);
+    }
+
+    public static function getJSONFromResponseBody($responseBody) {
+        error_log("ok json deserializing bdy");
```
Comment (matt-thomason) [inline Helper/Shared/ApiUtils.php]: This looks like a debugging statement.

File: Helper/Shared/ApiUtils.php [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        $additionalHeaders) {
+
+        return [
+                'User-Agent'            => 'BoltPay/Magento-'.$storeVersion . '/' . $moduleVersion,
+                'X-Bolt-Plugin-Version' => $moduleVersion,
+                'Content-Type'          => 'application/json',
+                'Content-Length'        => $requestData ? strlen($requestData) : null,
+                'X-Api-Key'             => $apiKey,
+                'X-Nonce'               => rand(100000000000, 999999999999)
+            ] + $additionalHeaders;
```
Comment (matt-thomason) [inline Helper/Shared/ApiUtils.php]: This is fine, but be aware that if `$additionalHeaders` has anything that over-writes the left hand array, the left hand array values win out.

File: Helper/GraphQL/Client.php [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+        } catch (\Exception $e) {
+            throw new LocalizedException($this->wrapGatewayError($e->getMessage()));
+        }
+
+        if ($responseBody) {
+            try {
+                $resultFromJSON = ApiUtils::getJSONFromResponseBody($responseBody);
+            } catch (\Exception $e) {
+                error_log($e);
+                error_log("all broke");
```
Comment (matt-thomason) [inline Helper/GraphQL/Client.php]: Second log seems unnecessary.

File: Helper/Shared/ApiUtils.php (hunk 11 above)
Comment (roopakv) [reply Helper/Shared/ApiUtils.php]: i just moved what we currently have. lmk if you think we should change the logic

**Comment by roopakv** [comment]
100% agree on unit tests, just managed to get it working. i couldnt figure it out yday. ill add more tests today, thank you all for the thorough and quick reviews (AND helping me learn)

File: Helper/Shared/ApiUtils.php (hunk 11 above)
Comment (matt-thomason) [reply Helper/Shared/ApiUtils.php]: s'all good.

**Comment by roopakv** [comment]
@daisy1754 @matt-thomason added some tests. This is mostly based on looking at phpunit docs since we dont have any complex tests which actually test our API. 

Let me know what you think.

File: Test/Unit/Helper/GraphQL/ClientTest.php [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+            ->willReturn($response);
+
+        $response->method('getBody')
+            ->willReturn('{"data": {"features": {"name": "OK", "value": true, "default_value": false, "rollout_percentage": 100}}}');
+
+        $boltResponseObserver = new ResponseObserver();
+
+        $this->responseFactory->method('create')->willReturn($boltResponseObserver);
+
+        $this->graphQLClient->getFeatureSwitches();
```
Comment (daisy1754) [inline Test/Unit/Helper/GraphQL/ClientTest.php]: can we assert return value instead?

File: Helper/GraphQL/Client.php [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+        $gqlRequest = array(
+            "operationName" => $operation,
+            "variables" => $variables,
+            "query" => $query
+        );
+
+        $requestData = json_encode($gqlRequest, JSON_UNESCAPED_SLASHES);
+
+        $apiURL = $this->configHelper->getApiUrl() . self::MERCHANT_API_GQL_ENDPOINT;
+        error_log($apiURL);
```
Comment (daisy1754) [inline Helper/GraphQL/Client.php]: let's remove this before merge

File: Helper/GraphQL/Client.php [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+            }
+
+            $result->setResponse($resultFromJSON);
+        } else {
+            throw new LocalizedException(__('Something went wrong when talking to Bolt.'));
+        }
+        return $result;
+    }
+
+    public function getFeatureSwitches() {
```
Comment (daisy1754) [inline Helper/GraphQL/Client.php]: let's add PHP doc

File: Test/Unit/Helper/GraphQL/ClientTest.php [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+
+        $httpClient->method('request')
+            ->willReturn($response);
+
+        $response->method('getBody')
+            ->willReturn('{"data": {"features": {"name": "OK", "value": true, "default_value": false, "rollout_percentage": 100}}}');
+
+        $boltResponseObserver = new ResponseObserver();
+
+        $this->responseFactory->method('create')->willReturn($boltResponseObserver);
```
Comment (daisy1754) [inline Test/Unit/Helper/GraphQL/ClientTest.php]: I think you can use Bolt\Boltpay\Model\Response here

File: Test/Unit/Helper/GraphQL/ResponseObserver.php [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+ * @category   Bolt
+ * @package    Bolt_Boltpay
+ * @copyright  Copyright (c) 2018 Bolt Financial, Inc (https://www.bolt.com)
+ * @license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+
+namespace Bolt\Boltpay\Test\Unit\Helper\GraphQL;
+
+
+class ResponseObserver {
```
Comment (daisy1754) [inline Test/Unit/Helper/GraphQL/ResponseObserver.php]: you can use Bolt\Boltpay\Model\Response instead

Comment (roopakv) [reply Test/Unit/Helper/GraphQL/ResponseObserver.php]: Yes but if I use response I am unable to store the the  response

Comment (roopakv) [reply Test/Unit/Helper/GraphQL/ResponseObserver.php]: As in I don't know how to test it

Comment (roopakv) [reply Test/Unit/Helper/GraphQL/ResponseObserver.php]: nvm got it

File: Helper/GraphQL/Client.php [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+            });
+
+            $this->bugsnag->registerCallback(function ($report) use ($response) {
+                $headers = $response->getHeaders();
+                $report->setMetaData([
+                    'META DATA' => [
+                        'bolt_trace_id' => @$headers[ConfigHelper::BOLT_TRACE_ID_HEADER],
+                    ]
+                ]);
+            });
```
Comment (danac-gs) [inline Helper/GraphQL/Client.php]: You can probably join these two callbacks into one.

Comment (roopakv) [reply Helper/GraphQL/Client.php]: one uses $request adn the the other $response? so maybe not. re-open if you feel strongly. Also i want to do one more refactoring pass but will do it once i get this in.

**Pull request merged**